components/eamxx/src/physics/mam/eamxx_mam_aci_functions.hpp

```diff
@@ -0,0 +1,726 @@
+      team_policy, KOKKOS_LAMBDA(const haero::ThreadTeam &team) {
+        const int icol = team.league_rank();
+        // for (int icol=0; icol<5; ++icol){
+        mam4::ndrop::View1D raercol_cw_view[mam4::ndrop::pver][2];
+        mam4::ndrop::View1D raercol_view[mam4::ndrop::pver][2];
+        for(int i = 0; i < mam4::ndrop::pver; ++i) {
+          for(int j = 0; j < 2; ++j) {
+            raercol_cw_view[i][j] = ekat::subview(loc_raercol_cw[i][j], icol);
+            raercol_view[i][j]    = ekat::subview(loc_raercol[i][j], icol);
+          }
+        }
+        mam4::ColumnView qqcw_view[mam4::ndrop::ncnst_tot];
+        for(int i = 0; i < mam4::ndrop::ncnst_tot; ++i) {
+          qqcw_view[i] = ekat::subview(qqcw_fld_work_loc[i], icol);
+        }
+        mam4::ColumnView ptend_q_view[mam4::aero_model::pcnst];
+        for(int i = 0; i < mam4::aero_model::pcnst; ++i) {
+          ptend_q_view[i] = ekat::subview(loc_ptend_q[i], icol);
+        }
+        mam4::ColumnView coltend_view[mam4::ndrop::ncnst_tot],
+            coltend_cw_view[mam4::ndrop::ncnst_tot];
+        for(int i = 0; i < mam4::ndrop::ncnst_tot; ++i) {
+          coltend_view[i]    = ekat::subview(loc_coltend[i], icol);
+          coltend_cw_view[i] = ekat::subview(loc_coltend_cw[i], icol);
+        }
+
+        // To construct state_q, we need fields from Prognostics (aerosols)
+        //  and Atmosphere (water species such as qv, qc etc.)
+
+        // get prognostics
+        mam4::Prognostics progs_at_col = aerosols_for_column(dry_aero, icol);
+
+        // get atmospheric quantities
+        haero::Atmosphere haero_atm =
+            atmosphere_for_column(dry_atmosphere, icol);
+
+        // Construct state_q (interstitial) and qqcw (cloud borne) arrays
+        Kokkos::parallel_for(
+            Kokkos::TeamThreadRange(team, 0u, mam4::ndrop::pver),
+            [=](int klev) {
+              Real state_q_at_lev_col[mam4::aero_model::pcnst] = {};
+
+              // get state_q at a grid cell (col,lev)
+              // NOTE: The order of species in state_q_at_lev_col
+              // is the same as in E3SM state%q array
+              mam4::utils::extract_stateq_from_prognostics(
+                  progs_at_col, haero_atm, state_q_at_lev_col, klev);
+
+              // get the start index for aerosols species in the state_q array
+              int istart = mam4::aero_model::pcnst - mam4::ndrop::ncnst_tot;
+
+              // create colum views of state_q
+              for(int icnst = istart; icnst < mam4::aero_model::pcnst;
+                  ++icnst) {
+                state_q_work_loc(icol, klev, icnst) = state_q_at_lev_col[icnst];
+              }
+
+              // get qqcw at a grid cell (col,lev)
+              // NOTE: The layout for qqcw array is based on mam_idx in
+              // dropmixnuc. To mimic that, we are using the following for-loops
+              int ind_qqcw = 0;
+              for(int m = 0; m < mam_coupling::num_aero_modes(); ++m) {
```

Comment:
These for loops will deteriorate performance on GPU, due to non-coalesced access. You should try to rework the code, with multiple TeamVectorRange loops (that's fine, especially if there are no deps among them, so no barriers needed). E.g., this for loop could be a TVR over nlevs*num_aero_mods, where inside you do modulo arithmetic to recover ilev/imode (in a way that still ensures coalesced access)

Reply:
Thanks. We will fix it in a follow-up PR (see issue #2858 )